### ISSUE TITLE
Wrong lexing of VimScript

### ISSUE THREAD

**Issue by sztomi** [isagalaev/highlight.js#1050]
This snippet:

```vim
" GoTo*
map <silent> <M-g> <leader>rj 
" Find references
map <silent> <M-S-f> <leader>rf
" Rename
map <silent> <M-S-r> <leader>rw
```

Is rendered like this:

![](http://i.imgur.com/BAEyvc3.png)

* The comments are lexed as strings (This is the offending line, but I couldn't fix it: https://github.com/isagalaev/highlight.js/blob/master/src/languages/vim.js#L49). Even vim fails highlighting comments properly sometimes, but I think this can be made more accurate. `^".*$` should definitely be a comment for example.
* `map` is lexed as a builtin, but it's a keyword
* g, f, r are lexed as keywords, but they aren't (I think they aren't ever keywords, just modifiers in commands?)

Possible improvements: `'strings'` also exist in VimScript. Also, the `<thingies>` are key bindings and could possible use a separate style.

**Comment by isagalaev** [comment]
Could you explain further about strings vs. comments? From what you said I can gather this:

- a double-quote at the beginning of the line starts a line comment
- a double-quote somewhere else is a string and can be ended with another double-quote or a new-line? or just a double-quote?

**Comment by isagalaev** [comment]
Also, is it right that `<...>` can only contain `\w+`? Or whitespace is allowed too?

**Comment by sztomi** [comment]
Thanks for the quick response. From my experience:

> a double-quote at the beginning of the line starts a line comment

Correct.

> a double-quote somewhere else is a string and can be ended with another double-quote or a new-line? or just a double-quote?

In my experience, it is like this:

* double quote in the middle of the line + content + newline = comment. Mind you, this is something that vim fails to highlight properly:
![](http://i.imgur.com/cml649C.png)

* double quote in the middle of the line + content + newline + \ at the beginning of the next line = string

> Also, is it right that <...> can only contain \w+? Or whitespace is allowed too?

No, I think it is closer to `[\w\d\-]+`.

I'm not terribly knowledgable about VimScript though, but the above rules definitely give a better approximation.

**Comment by isagalaev** [comment]
Ok, got it.

> double quote in the middle of the line + content + newline + \ at the beginning of the next line = string

That's crazy :-) I'll see what I can do!

**Comment by sztomi** [comment]
> That's crazy :-)

That's VimScript in a nutshell :) The *leading* backslash is the line continuation symbol in the language to avoid some incompatibility. Here's the docs: http://vimdoc.sourceforge.net/htmldoc/repeat.html#line-continuation

**Comment by isagalaev** [comment]
Work in progress: 
![screenshot from 2016-01-12 15-46-26](https://cloud.githubusercontent.com/assets/99931/12280580/da8c70da-b943-11e5-9016-29119d28370f.png)

**Comment by isagalaev** [comment]
Okay, I've fixed it to the best of my abilities:

- Key bindings (`<..>`) are now recognized as symbols
- Comments and strings are mostly correct (see note at the end of f2e6828b0)
- Removed built-ins that were also listed as keywords (including 'map')

I don't really know what to do with those single-letter keywords, I don't use VimScript at all. @sztomi if you feel they'd rather go away, feel free to file a pull request with this change. There are also suspiciously named built-ins: "and", "or". May be they're keywords too?

**Comment by sztomi** [comment]
Wow, that was fast. Thanks! How stable is HEAD? Can I just grab the current commit and use it on my blog? (I'm writing a blog post and I encountered the above bug during that). Or is there a release coming soon?

> I don't really know what to do with those single-letter keywords, I don't use VimScript at all. @sztomi if you feel they'd rather go away, feel free to file a pull request with this change. 

Will do. At best, they are Ex "keywords" but I'm not sure if there is such terminology for Ex even. Vim doesn't highlight them.

> There are also suspiciously named built-ins: "and", "or". May be they're keywords too?

They are definitely not keywords, but I'm not sure if they even exist in vanilla vim. I found [this](https://neovim.io/doc/user/usr_41.html), which states they are bitwise functions, but I couldn't find them in the vim version of the same document. I also found some mailing list threads where vim authors were reluctant to add them. Anyway, neither vim nor neovim highlights these (or any builtins, as far as I can tell).

**Comment by isagalaev** [comment]
> Wow, that was fast.

To be honest, I started to tackle it only because I thought it'd be a very quick fix. And when it became apparent it wasn't, it was too late to just drop it on the floor, so I had to finish it :-)

> How stable is HEAD? Can I just grab the current commit and use it on my blog?

Pretty stable. If the tests pass, the build is highly unlikely to hang anything. Also, we're just past a big naming change of CSS classes, so you won't even be affected by this once-in-a-few-years breakage.

> Or is there a release coming soon?

We do it regularly every 6 weeks. The next one is planned on Feb, 23.

> Anyway, neither vim nor neovim highlights these (or any builtins, as far as I can tell). 

Let them live for now then. No harm in having those, I think.
